storagenode/piecestore/endpoint.go

```diff
@@ -314,7 +317,19 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 	largestOrder := pb.Order{}
 	defer commitOrderToStore(ctx, &largestOrder)
 
+	// monitor speed of upload client to flag out slow uploads.
+	speedEstimate := speedEstimation{
+		grace: endpoint.config.MinUploadSpeedGraceDuration,
+		limit: endpoint.config.MinUploadSpeed,
+	}
+
 	for {
+		congested := endpoint.isCongested()
+
+		if err := speedEstimate.EnsureLimit(memory.Size(pieceWriter.Size()), congested, time.Now()); err != nil {
```

Comment:
NIT: This can be combined into one line and avoids the var assignment.

Reply:
done